### PR TITLE
gguf : align the data section relative to the GGUF start, not the file

### PR DESCRIPTION
**Pull request by ykhrustalev**
## Overview

Rework of #28973 using `FILE *` instead of file descriptors, as asked there and in #20402.
#28973 can be closed.

`gguf_init_from_file_ptr()` reads from the current file position, but aligns the data
section from offset 0. So a GGUF stored at an unaligned offset inside a bigger file (an
Android APK asset, for example) loads without error and returns wrong tensor data. The
reader now aligns from where the GGUF starts.

With that fixed, no new API is needed for embedded models: `fseek` to the GGUF, then
`llama_model_load_from_file_ptr()`. On Android:

```c
FILE * f = fdopen(dup(fd), "rb");      // fd, start, length from AssetFileDescriptor
fseeko(f, start, SEEK_SET);
llama_model * model = llama_model_load_from_file_ptr(f, params);
fclose(f);
```

Two small additions:

- `llama_adapter_lora_init_from_file_ptr()`, since LoRA had no `FILE *` entry point.
- If the GGUF data is not on a 32 byte file offset, mmap is turned off with a warning
  instead of ggml asserting.

## Additional information

- `test-gguf`: new `file_offset` mode, GGUF written after 7 junk bytes. Fails without
  the fix.
- `test-load-file-ptr` (needs the model fixture): model and LoRA loaded from an offset,
  logits compared with a normal load. It is a new test file; drop it if not wanted.
- Passes on macOS arm64 and a Galaxy S24 Ultra, and from a real APK on the S24 where the
  LoRA asset landed at offset mod 32 = 16.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES. An assistant helped find the bug, write the fix and tests, and
  run the checks.

### CODE REVIEW

File: ggml/src/gguf.cpp [hunk 1]
```diff
@@ -366,6 +367,11 @@ struct gguf_reader {
         return data_offset;
     }
 
+    // position where the GGUF starts, alignment is relative to it, not to the file
```
Comment (JohannesGaessler) [inline ggml/src/gguf.cpp]:
```suggestion
    // position in the file where the GGUF data starts, alignment is relative to it, not to the file
```

File: src/llama-model-loader.cpp [hunk 2]
```diff
@@ -685,6 +685,14 @@ llama_model_loader::llama_model_loader(
             throw std::runtime_error(format("%s: failed to load model from file pointer", __func__));
         }
 
+        // mmap places tensors at their file offsets, so an embedded GGUF must be aligned in the file too
+        const size_t tensor_align = ggml_backend_buft_get_alignment(ggml_backend_cpu_buffer_type());
+        if (use_mmap && gguf_get_data_offset(metadata) % tensor_align != 0) {
+            LLAMA_LOG_WARN("%s: GGUF data section at file offset %zu is not %zu byte aligned, mmap is disabled\n",
+                __func__, gguf_get_data_offset(metadata), tensor_align);
+            use_mmap = false;
+        }
```
Comment (JohannesGaessler) [inline src/llama-model-loader.cpp:693]: This should throw a runtime error instead.

File: include/llama.h [hunk 3]
```diff
@@ -518,6 +518,8 @@ extern "C" {
               struct llama_model_params   params);
 
     // Load a model from an open FILE pointer
+    // The GGUF is read from the current position, so it can be embedded in a larger file
+    // mmap needs the GGUF data section at a file offset aligned to the CPU tensor alignment (32 bytes), otherwise it is disabled
```
Comment (JohannesGaessler) [inline include/llama.h]:
```suggestion
    // mmap needs the GGUF data section at a file offset to be aligned to the CPU tensor alignment (32 bytes)
```

File: tests/test-load-file-ptr.cpp
Comment (JohannesGaessler) [inline tests/test-load-file-ptr.cpp:1]: My opinion is that this test should be removed because maintaining it will be more maintenance effort than it saves through its existence.

Comment (ykhrustalev) [reply tests/test-load-file-ptr.cpp:1]: I won't object, added it for a demo case

**Review by JohannesGaessler — APPROVED** [review]

**Review by CISC — APPROVED** [review]
